Mathlib/Analysis/NormedSpace/Alternating.lean

```diff
@@ -0,0 +1,1150 @@
+lemma op_norm_le_bound {M : ℝ} (hMp : 0 ≤ M) (hM : ∀ m, ‖f m‖ ≤ M * ∏ i, ‖m i‖) :
+    ‖f‖ ≤ M :=
+  f.1.op_norm_le_bound hMp hM
+
+section
+
+variable {𝕜' : Type*} [NormedField 𝕜'] [NormedSpace 𝕜' G] [SMulCommClass 𝕜 𝕜' G]
+
+instance instNormedSpace : NormedSpace 𝕜' (E [Λ^ι]→L[𝕜] G) :=
+  ⟨fun c f ↦ f.1.op_norm_smul_le c⟩
+
+variable (𝕜')
+
+/-- The inclusion of *alternating* continuous multi-linear maps into continuous multi-linear maps
+as a continuous linear map. -/
+@[simps!]
+def toContinuousMultilinearMapL :
+    E [Λ^ι]→L[𝕜] G →L[𝕜'] ContinuousMultilinearMap 𝕜 (fun _ : ι ↦ E) G :=
+  ⟨toContinuousMultilinearMapLinear, continuous_induced_dom⟩
+
+variable {𝕜'}
+
+theorem le_op_norm_mul_prod_of_le {b : ι → ℝ} (hm : ∀ i, ‖m i‖ ≤ b i) : ‖f m‖ ≤ ‖f‖ * ∏ i, b i :=
+  f.1.le_op_norm_mul_prod_of_le hm
+
+theorem le_op_norm_mul_pow_card_of_le {b : ℝ} (hm : ‖m‖ ≤ b) : ‖f m‖ ≤ ‖f‖ * b ^ Fintype.card ι :=
+  f.1.le_op_norm_mul_pow_card_of_le hm
+
+theorem le_op_norm_mul_pow_of_le (f : E [Λ^Fin n]→L[𝕜] G) (m : Fin n → E) {b : ℝ} (hm : ‖m‖ ≤ b) :
+    ‖f m‖ ≤ ‖f‖ * b ^ n :=
+  f.1.le_op_norm_mul_pow_of_le hm
+
+/-- The fundamental property of the operator norm of a continuous alternating map:
+`‖f m‖` is bounded by `‖f‖` times the product of the `‖m i‖`, `nnnorm` version. -/
+theorem le_op_nnnorm : ‖f m‖₊ ≤ ‖f‖₊ * ∏ i, ‖m i‖₊ := f.1.le_op_nnnorm m
+
+theorem le_of_op_nnnorm_le {C : ℝ≥0} (h : ‖f‖₊ ≤ C) : ‖f m‖₊ ≤ C * ∏ i, ‖m i‖₊ :=
+  f.1.le_of_op_nnnorm_le m h
+
+lemma op_norm_prod (f : E [Λ^ι]→L[𝕜] G) (g : ContinuousAlternatingMap 𝕜 E G' ι) :
+    ‖f.prod g‖ = max (‖f‖) (‖g‖) :=
+  f.1.op_norm_prod g.1
+
+lemma op_norm_pi {ι' : Type v'} [Fintype ι'] {E' : ι' → Type wE'} [∀ i', NormedAddCommGroup (E' i')]
+    [∀ i', NormedSpace 𝕜 (E' i')] (f : ∀ i', ContinuousAlternatingMap 𝕜 E (E' i') ι) :
+    ‖pi f‖ = ‖f‖ :=
+  ContinuousMultilinearMap.op_norm_pi fun i ↦ (f i).1
+
+section
+variable (𝕜 G)
+
+lemma norm_of_subsingleton_le [Subsingleton ι] (i' : ι) : ‖ofSubsingleton 𝕜 G i'‖ ≤ 1 :=
+  ContinuousMultilinearMap.norm_ofSubsingleton_le 𝕜 G i'
+
+@[simp] lemma norm_of_subsingleton [Subsingleton ι] [Nontrivial G] (i' : ι) :
+    ‖ofSubsingleton 𝕜 G i'‖ = 1 :=
+  ContinuousMultilinearMap.norm_ofSubsingleton 𝕜 G i'
+
+lemma nnnorm_of_subsingleton_le [Subsingleton ι] (i' : ι) : ‖ofSubsingleton 𝕜 G i'‖₊ ≤ 1 :=
+  norm_of_subsingleton_le _ _ _
+
+@[simp] lemma nnnorm_of_subsingleton [Subsingleton ι] [Nontrivial G] (i' : ι) :
+    ‖ofSubsingleton 𝕜 G i'‖₊ = 1 :=
+  NNReal.eq <| norm_of_subsingleton _ _ _
+
+variable {G} (E)
+
+@[simp] lemma norm_constOfIsEmpty [IsEmpty ι] (x : G) : ‖constOfIsEmpty 𝕜 E ι x‖ = ‖x‖ :=
+  ContinuousMultilinearMap.norm_constOfIsEmpty _ _ _
+
+@[simp] lemma nnnorm_constOfIsEmpty [IsEmpty ι] (x : G) : ‖constOfIsEmpty 𝕜 E ι x‖₊ = ‖x‖₊ :=
+  NNReal.eq <| norm_constOfIsEmpty _ _ _
+
+end
+
+section
+
+variable (𝕜 E E' G G')
+
+/-- `ContinuousMultilinearMap.prod` as a `LinearIsometryEquiv`. -/
```

Comment:
This docstring looks wrong

Reply:
And not only this one... Converting to draft.